<a href="https://colab.research.google.com/github/Rochy3/AlgorithmsUN2024II/blob/main/Lab3/RosaSantofimio_Lab3.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [11]:
!cat /proc/cpuinfo

processor	: 0
vendor_id	: GenuineIntel
cpu family	: 6
model		: 79
model name	: Intel(R) Xeon(R) CPU @ 2.20GHz
stepping	: 0
microcode	: 0xffffffff
cpu MHz		: 2199.998
cache size	: 56320 KB
physical id	: 0
siblings	: 2
core id		: 0
cpu cores	: 1
apicid		: 0
initial apicid	: 0
fpu		: yes
fpu_exception	: yes
cpuid level	: 13
wp		: yes
flags		: fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush mmx fxsr sse sse2 ss ht syscall nx pdpe1gb rdtscp lm constant_tsc rep_good nopl xtopology nonstop_tsc cpuid tsc_known_freq pni pclmulqdq ssse3 fma cx16 pcid sse4_1 sse4_2 x2apic movbe popcnt aes xsave avx f16c rdrand hypervisor lahf_lm abm 3dnowprefetch invpcid_single ssbd ibrs ibpb stibp fsgsbase tsc_adjust bmi1 hle avx2 smep bmi2 erms invpcid rtm rdseed adx smap xsaveopt arat md_clear arch_capabilities
bugs		: cpu_meltdown spectre_v1 spectre_v2 spec_store_bypass l1tf mds swapgs taa mmio_stale_data retbleed bhi
bogomips	: 4399.99
clflush size	: 64
cache_alignment	: 64
ad

# I Solving the Sum of Two Digits Programming Challenge

## C++

In [12]:
%%writefile APlusB.cpp
#include <iostream>

int sum_of_two_digits(int first_digit, int second_digit) {
    return first_digit + second_digit;
}

int main() {
    int a = 0;
    int b = 0;
    std::cin >> a;
    std::cin >> b;
    std::cout << sum_of_two_digits(a, b);
    return 0;
}

Writing APlusB.cpp


In [13]:
!g++ APlusB.cpp -o APlusB -std=c++11
!./APlusB

2
3
5

### Test input output files

In [14]:
%%writefile input1.txt
4 8

Writing input1.txt


In [15]:
%%writefile output1.txt
12

Writing output1.txt


In [16]:
%%writefile input2.txt
0 9

Writing input2.txt


In [17]:
%%writefile output2.txt
9

Writing output2.txt


In [18]:
!./APlusB < input1.txt

12

In [19]:
!./APlusB < input2.txt

9

### Comparing files with grep

https://askubuntu.com/questions/546796/comparing-two-text-files

In [20]:
!grep -xvFf output1.txt output1.txt

In [21]:
!grep -xvFf output1.txt output2.txt

9


In [22]:
!grep -xvFf output2.txt output1.txt

12


In [23]:
!./APlusB < input1.txt> ctest1.txt
!grep -xvFf output1.txt ctest1.txt

In [24]:
!./APlusB < input2.txt > ctest2.txt
!grep -xvFf output2.txt ctest2.txt

## Python

In [25]:
%%writefile sum.py
# python3

def sum_of_two_digits(first_digit, second_digit):
    return first_digit + second_digit

if __name__ == '__main__':
    a, b = map(int, input().split())
    print(sum_of_two_digits(a, b))

Writing sum.py


In [26]:
!python sum.py

5
Traceback (most recent call last):
  File "/content/sum.py", line 7, in <module>
    a, b = map(int, input().split())
ValueError: not enough values to unpack (expected 2, got 1)


In [27]:
!python sum.py  < input1.txt> ptest1.txt
!grep -xvFf output1.txt ptest1.txt

In [28]:
!python sum.py  < input2.txt> ptest2.txt
!grep -xvFf output2.txt ptest2.txt

# II Solving The Maximum Pairwise Product Programming Challenge in C++

## C++

$$ T(n) < 6 n^2 = O(n^2)$$

In [29]:
%%writefile max_pairwise_product_full.cpp
#include <iostream>
#include <vector>
#include <algorithm>

int MaxPairwiseProduct(const std::vector<int>& numbers) {
    int max_product = 0;
    int aux_product = 0;
    int n = numbers.size();

    for (int first = 0; first < n; ++first) {
        for (int second = 0; second < n; ++second) {
          if ( first != second ){
                aux_product = numbers[first] * numbers[second];
                if (aux_product > max_product)  max_product = aux_product;
          }
        }
    }

    return max_product;
}

int main() {
    int n;
    std::cin >> n;
    std::vector<int> numbers(n);
    for (int i = 0; i < n; ++i) {
        std::cin >> numbers[i];
    }

    std::cout << MaxPairwiseProduct(numbers) << "\n";
    return 0;
}

Writing max_pairwise_product_full.cpp


#### Time Complexity
$$\left. \begin{array}{llc}
 \text{INSTRUCTION} & \text{TYPE}& \text{# EXECUTIONS}\\
\text{first != second} &  \text{comparisons} &n^2 \\
\text{numbers[first] * numbers[second]}& \ \text{ products}  &n^2 - n \\
\text{max_product = aux_product} &  \text{assigments} & [1, \frac{n^2 - n}{2})  
\end{array}\right\}  =  O(n^2)$$

In [30]:
!g++ max_pairwise_product_full.cpp -o max_pairwise_product_full
!./max_pairwise_product_full

2
3
2
6


### Test 1: Small numbers

In [31]:
%%writefile input1.txt
10
1 2 3 4 5 6 7 8 9 10

Overwriting input1.txt


In [32]:
!./max_pairwise_product < input1.txt

/bin/bash: line 1: ./max_pairwise_product: No such file or directory


In [33]:
%%writefile output1.txt
90

Overwriting output1.txt


In [34]:
!./max_pairwise_product_full< input1.txt > ctest1.txt
!grep -xvFf output1.txt ctest1.txt

In [35]:
%%writefile input2.txt
10
10 9 8 7 6 5 4 3 2 1

Overwriting input2.txt


In [36]:
%%writefile output2.txt
90

Overwriting output2.txt


In [37]:
!./max_pairwise_product_full< input2.txt > ctest2.txt
!grep -xvFf output2.txt ctest2.txt

### Test 2: largest input numbers 200000 200000 to check possible intger overflow

In [38]:
%%writefile input3.txt
6
1 3 200000 9 200000 23

Writing input3.txt


In [39]:
200000*200000

40000000000

In [40]:
%%writefile output3.txt
40000000000

Writing output3.txt


In [41]:
!./max_pairwise_product_full < input3.txt

1345294336


In [42]:
!./max_pairwise_product_full < input3.txt > ctest3.txt
!grep -xvFf output3.txt ctest3.txt

1345294336


In [43]:
 %%writefile run_all_test.sh
#!/bin/bash
g++ max_pairwise_product_full.cpp -o max_pairwise_product_full
for i in {0..3}
do
    ./max_pairwise_product_full < input$i.txt > ctest$i.txt
    grep -xvFf output$i.txt ctest$i.txt
done

Writing run_all_test.sh


In [44]:
!chmod u+x run_all_test.sh
!./run_all_test.sh

./run_all_test.sh: line 5: input0.txt: No such file or directory
grep: output0.txt: No such file or directory
1345294336


### Changing the variables to long long

In [45]:
%%writefile max_pairwise_product_long_full.cpp
#include <iostream>
#include <vector>
#include <algorithm>

long long MaxPairwiseProduct(const std::vector<int>& numbers) {
    long long max_product = 0LL;
	  long long aux_product = 0LL;
    int n = numbers.size();

    for (int first = 0; first < n; ++first) {
        for (int second = 0; second < n; ++second) {
          if ( first != second){
                aux_product = (long long) numbers[first] * numbers[second];
                if (aux_product > max_product)  max_product = aux_product;
          }
        }
    }


    return max_product;
}

int main() {
    int n;
    std::cin >> n;
    std::vector<int> numbers(n);
    for (int i = 0; i < n; ++i) {
        std::cin >> numbers[i];
    }

    std::cout << MaxPairwiseProduct(numbers) << "\n";
    return 0;
}


Overwriting max_pairwise_product_long_full.cpp


In [46]:
!g++ max_pairwise_product_long_full.cpp -o max_pairwise_product_long_full

In [47]:
!./max_pairwise_product_long_full < input3.txt

40000000000


### Runing multuple test with a bash shell

In [48]:
 %%writefile run_all_test.sh
#!/bin/bash
g++ max_pairwise_product_long_full.cpp -o max_pairwise_product_long_full
for i in {1..3}
do
    ./max_pairwise_product_long_full < input$i.txt > ctest$i.txt
    grep -xvFf output$i.txt ctest$i.txt
done

Overwriting run_all_test.sh


In [49]:
!chmod u+x run_all_test.sh
!./run_all_test.sh

### Test 3  stress tests, 6 testing cases files input4.txt/output4.txt .. input10.txt/output10.txt

---



In [50]:

%%writefile input4.txt
7
1 3 200000 9 200000 23 200000

Writing input4.txt


In [51]:

%%writefile output4.txt
40000000000

Writing output4.txt


In [52]:

%%writefile input5.txt
2
1 1

Writing input5.txt


In [53]:

%%writefile output5.txt
1

Writing output5.txt


In [54]:
%%writefile input6.txt
8
2 2 2 2 2 2 2 2

Writing input6.txt


In [55]:
%%writefile output6.txt
4

Writing output6.txt


In [56]:
%%writefile input7.txt
5
1 3 200000 9 200000


Writing input7.txt


In [57]:
%%writefile output7.txt
40000000000

Writing output7.txt


In [58]:
%%writefile input8.txt
3
9 8 9

Writing input8.txt


In [59]:
%%writefile output8.txt
81

Writing output8.txt


In [60]:

%%writefile input9.txt
4
200000 200000 200000 200000

Writing input9.txt


In [61]:
%%writefile output9.txt
40000000000

Writing output9.txt


In [62]:
%%writefile input10.txt
3
0 0 0

Writing input10.txt


In [63]:

%%writefile output10.txt
0


Writing output10.txt


In [64]:
%%writefile run_all_test_1.sh
#!/bin/bash
g++ max_pairwise_product_long_full.cpp -o max_pairwise_product_long_full || exit 1

for i in {4..10}
do
    ./max_pairwise_product_long_full < input$i.txt > ctest$i.txt
    if grep -xvFf output$i.txt ctest$i.txt; then
        echo "Test $i failed!"
    else
        echo "Test $i passed!"
    fi
done


Writing run_all_test_1.sh


In [65]:
!chmod u+x run_all_test_1.sh
!./run_all_test_1.sh

Test 4 passed!
Test 5 passed!
Test 6 passed!
Test 7 passed!
Test 8 passed!
Test 9 passed!
Test 10 passed!


### Test 4  time test  measure the execution time running with the the largest worst cases input11.txt  (200000 increasing numbers) and input12.txt (199998 increasing numbers and  200000, 200000 at the end)

In [66]:
# input11.txt 200000 increasing numbers:  1 2 .. 200000

%%writefile WriteTest.cpp

#include <iostream>
#include <fstream>
#include <cstdlib>

using namespace std;

int main() {
  // Create and open a text file
  ofstream MyFile("input11.txt");

 MyFile << 200000 <<"\n";  // n
 for (int i=1; i <200001; i++){
     MyFile << i <<" ";
 }

  // Close the file
  MyFile.close();
}

Writing WriteTest.cpp


In [67]:
!g++ WriteTest.cpp -o WriteTest
!./WriteTest

In [68]:
199999 * 200000

39999800000

In [69]:
%%writefile output11.txt
39999800000

Writing output11.txt


In [70]:
# input12.txt 200000 decreasing numbers and 200000, 199999 ...  2 1

%%writefile WriteTest.cpp

#include <iostream>
#include <fstream>
#include <cstdlib>

using namespace std;

int main() {
  // Create and open a text file
  ofstream MyFile("input12.txt");

 MyFile << 200000 <<"\n";  // n
 for (int i=1; i <200001; i++){
     MyFile << 200001-i <<" ";
 }

  // Close the file
  MyFile.close();
}

Overwriting WriteTest.cpp


In [71]:
!g++ WriteTest.cpp -o WriteTest
!./WriteTest

In [72]:
%%writefile output12.txt
39999800000

Writing output12.txt


### Measure time

https://askubuntu.com/questions/53444/how-can-i-measure-the-execution-time-of-a-terminal-process

In [114]:
!g++ max_pairwise_product_long_full.cpp -o max_pairwise_product_long_full
!time ./max_pairwise_product_long_full < input11.txt

^C


In [74]:
!g++ max_pairwise_product_long_full.cpp -o max_pairwise_product_long_full
!time ./max_pairwise_product_long_full< input12.txt

^C


### Calculate the number or and upper bound to the number of instructions executed by the algorithm as a function of n, add in each line a comment with the number of instructions that are executed in that line.

In [75]:
%%writefile max_pairwise_product_long_full.cpp
#include <iostream>
#include <vector>
#include <algorithm>

long long MaxPairwiseProduct(const std::vector<int>& numbers) {              //INSTRUCTIONS executed by line
    long long max_product = 0LL;                                             //1
	  long long aux_product = 0LL;                                             //1
    int n = numbers.size();                                                  //1

    for (int first = 0; first < n; ++first) {                                //2+2n
        for (int second = 0; second < n; ++second) {                         //2n + 2n**2
          if ( first != second ){                                            //n**2
                aux_product = numbers[first] * numbers[second];              //n**2 -n
                if (aux_product > max_product)                               //n**2 -n
                    max_product = aux_product;                               //[0, n**2/2 - n/2] upper bound
          }
        }
    }

                                                                             // --------------
    return max_product;                                                      // 5 + 2*n + [5*n**2, 11*n**2/2  - n/2]  upper bound
}

int main() {                                              //INSTRUCTIONS executed by line
    int n;                                                //1
    std::cin >> n;                                        //1
    std::vector<int> numbers(n);                          //n
    for (int i = 0; i < n; ++i) {                         //2 + 2n
        std::cin >> numbers[i];                           //n
    }

    std::cout << MaxPairwiseProduct(numbers) << "\n";     //1 + 1  + 5 + 2*n + [5*n**2, 11*n**2/2  - n/2]

    return 0;                                             // --------------
}                                                         //  11 + 6*n + [5*n**2, 11*n**2/2  - n/2]  upper bound


Overwriting max_pairwise_product_long_full.cpp


### Count executed intructions in code

In [76]:
%%writefile max_pairwise_product_long_full_execints.cpp
#include <iostream>
#include <vector>
#include <algorithm>

long long MaxPairwiseProduct(const std::vector<int>& numbers, long long &execints) {      //INSTRUCTIONS executed by line

    long long max_product = 0LL;                                                          //1
    execints++;

	  long long aux_product = 0LL;                                                          //1
    execints++;

    int n = numbers.size();                                                               //1
    execints++;

    execints++;
    for (int first = 0; first < n; ++first) {                                             //2+2n
        execints += 2;

        execints++;
        for (int second = 0; second < n; ++second) {                                       //2n + 2n**2
          execints += 2;

          execints++;
          if ( first != second ){                                                         //n**2
                aux_product = numbers[first] * numbers[second];                           //n**2 -n
                execints++;

                execints++;
                if (aux_product > max_product){                                           //n**2 -n
                    max_product = aux_product;                                            //[0,n**2/2 - n/2)  upper bound
                    execints++;

                }
        }

      }
      execints++;
    }
    execints++;
                                                                                        // --------------
    return max_product;                                                                 // 5 + 2*n + [5*n**2, 11*n**2/2  - n/2] upper bound
}

int main() {                                                 //INSTRUCTIONS executed by line

    long long execints =0LL;

    int n;                                                   //1
    execints++;

    std::cin >> n;                                           //1
    execints++;

    std::vector<int> numbers(n);                             //n
    execints += n;

    execints++;
    for (int i = 0; i < n; ++i) {                            //2 + 2n
        execints += 2;

        std::cin >> numbers[i];
        execints++;                                          //n

    }


    //std::cout << MaxPairwiseProduct(numbers) << "\n";      //1 + 1  + 5 + 2*n + [5*n**2, 11*n**2/2  - n/2] upper bound

    execints++;
    int M = MaxPairwiseProduct(numbers,execints);

    execints++;
    std::cout <<  execints << "\n";
    return 0;                                                // --------------
}                                                            // 11 + 6*n + [5*n**2, 11*n**2/2  - n/2]

Overwriting max_pairwise_product_long_full_execints.cpp


The number of instructions executed by the algooirithm with a list of size $n$ is  

$$T (n) = 11 + 6n + \left[5n^2, 11\frac{n^2}{2}  - \frac{n}{2}\right)  >  6n^2 = O(n^2)$$

It depends on the size of the lists and  the order of the list.

For the input0.txt
```
10
0 0 0 0 0 0 0 0 0 0
```
$$T = 11 + 6(10) + 5(10)^2   = 561 $$

$$T(10) \in [571, 621)  $$


In [77]:
!g++ max_pairwise_product_long_full_execints.cpp -o max_pairwise_product_long_full_execints
!./max_pairwise_product_long_full_execints < input0.txt

/bin/bash: line 1: input0.txt: No such file or directory


For the input1.txt
```
10
1 2 3 4 5 6 7 8 9 10
```
$$T = 11 + 6(10) + 5(10)^2 +  9 + 5 + 4 + 3 + 2 + 2 + 2 + 2 + 1 + 0  = 601 $$

$$T(10) \in [572, 622)  $$


In [78]:
#The number of times that  max_product = aux_product; in the worst case is less than (n**2 -n)/2 -  upper bound

#     1  2  3  4  5  6  7  8  9  10
#  1  x  2  3  4  5  6  7  8  9  10         9
#  2  2  x  6  8  10 12 14 16 18 20         5 14
#  3  3  6  x  12 15 18 21 24 27 30         4 18
#  4  4  8  12 x  20 24 28 32 36 40         3 21
#  5  5  10 15 20 x  30 35 40 45 50         2 23
#  6  6  12 18 24 30 x  42 48 54 60         2 25
#  7  7  14 21 28 35 42 x  56 63 70         2 27
#  8  8  18 27 36 45 54 63 72 x  90         1 30
#  10 10 20 30 40 50 60 70 60 90 x          0

In [79]:
11 + 6*(10) + 5*(10)*(10) +  9 + 5 + 4 + 3 + 2 + 2 + 2 + 2 + 1 + 0

601

In [80]:
n = 10
print( "[", 11 + 6*n + 5*n**2,  ",", 11 + 6*n  + 5.5*n**2, ")")

[ 571 , 621.0 )


In [81]:
!g++ max_pairwise_product_long_full_execints.cpp -o max_pairwise_product_long_full_execints
!./max_pairwise_product_long_full_execints < input1.txt

600



For the input2.txt
```
10
10 9 8 7 6 5 4 3 2 1
```
$$T = 11 + 6*(10) + 5*(10)**2 +  1  = 571 $$

In [82]:
11 + 6*(10) + 5*(10)**2 +  1

572

In [83]:
!g++ max_pairwise_product_long_full_execints.cpp -o max_pairwise_product_long_full_execints
!./max_pairwise_product_long_full_execints < input2.txt

571


Calculate estimated time by instruction and number of instructions by second

For the input11.txt
```
200000
1 2 ..  199999 2000
```
$$T = 11 + 6*(200000) + 5.5*(200000)**2  = 200001200012  $$

In [116]:

11 + 6*(200000) + 5.5*(200000)**2+1

220001200012.0

In [84]:
!g++ max_pairwise_product_long_full_execints.cpp -o max_pairwise_product_long_full_execints
!./max_pairwise_product_long_full_execints < input11.txt

^C


In [85]:
# Estimated time by instruction
# use the real reported  !time ./max_pairwise_product_long_full< input11.txt in seconds intead of the 295

295/200003177092


1.4749765693187072e-09

In [86]:
# Estimated number of instructions by second

200003177092/295

677976871.4983051

### Estimate and verify the maximun n that can be solved in 1 second

$$(5.5n^2 + 5.5n + 11) \times 1.47*10^{-9} s = 1 s$$

$$ n \approx 11000$$


In [87]:
# Verify input13.txt
%%writefile WriteTest.cpp

#include <iostream>
#include <fstream>
#include <cstdlib>

using namespace std;

int main() {
  // Create and open a text file
  ofstream MyFile("input13.txt");

 MyFile << 11000 <<"\n";  // n
 for (int i=1; i <11001; i++){
     MyFile << i <<" ";
 }

  // Close the file
  MyFile.close();
}

Overwriting WriteTest.cpp


In [88]:
!g++ WriteTest.cpp -o WriteTest
!./WriteTest

In [89]:
10999 * 11000

120989000

In [90]:
!g++ max_pairwise_product_long_full.cpp -o max_pairwise_product_long_full
!time ./max_pairwise_product_long_full< input13.txt

120989000

real	0m1.138s
user	0m1.104s
sys	0m0.005s


## Python

In [117]:

%%writefile max_pairwise_product_long_full.py
def MaxPairwiseProduct(numbers):
  max_product = 0
  aux_product = 0
  n = len(numbers)

  for first in range (n):
    for second in range(n):
      if(first != second):
        aux_product = numbers[first]*numbers[second]
        if aux_product > max_product:
          max_product = aux_product
  return max_product

n = int(input())
numbers = [int(i) for i in input().split()]
print(MaxPairwiseProduct(numbers))


Writing max_pairwise_product_long_full.py


In [118]:
# Verify
!time python3 max_pairwise_product_long_full.py < input13.txt

Traceback (most recent call last):
  File "/content/max_pairwise_product_long_full.py", line 16, in <module>
    print(MaxPairwiseProduct(numbers))
  File "/content/max_pairwise_product_long_full.py", line 10, in MaxPairwiseProduct
    if aux_product > max_product:
KeyboardInterrupt
^C


# For the follwing solutions rquired  the same tests and analysis performed for the first solution

## Solutions in C++ and Python that uses
$$\frac{n^2 - n}{2} = O(n^2)\ \ \text{products}$$

In [92]:
%%writefile max_pairwise_product.cpp
#include <iostream>
#include <vector>
#include <algorithm>

int MaxPairwiseProduct(const std::vector<int>& numbers) {
    int max_product = 0;
    int n = numbers.size();

    for (int first = 0; first < n; ++first) {
        for (int second = first + 1; second < n; ++second) {
            max_product = std::max(max_product,
                numbers[first] * numbers[second]);
        }
    }

    return max_product;
}

int main() {
    int n;
    std::cin >> n;
    std::vector<int> numbers(n);
    for (int i = 0; i < n; ++i) {
        std::cin >> numbers[i];
    }

    std::cout << MaxPairwiseProduct(numbers) << "\n";
    return 0;
}

Writing max_pairwise_product.cpp


In [93]:
%%writefile max_pairwise_product_long.cpp
#include <iostream>
#include <vector>
#include <algorithm>

long long MaxPairwiseProduct(const std::vector<int>& numbers) {              //INSTRUCTIONS executed by line
    long long max_product = 0LL;                                             //1
	  long long aux_product = 0LL;                                             //1
    int n = numbers.size();                                                  //1

    for (int first = 0; first < n; ++first) {                                //2+2n
        for (int second = first + 1; second < n; ++second) {                 //2n + (n-1)n
			       aux_product =  (long long) numbers[first] * numbers[second];    //(n-1)n/2
            if (max_product < aux_product)  {                                //(n-1)n/2
				        max_product = aux_product;                                   //[1,(n-1)n/2)  upper bound
             }
        }
    }
                                                                             // --------------
    return max_product;                                                      // 5 + 4n + [2(n-1)n+1, 2.5 (n-1)n) upper bound
}

int main() {                                              //INSTRUCTIONS executed by line
    int n;                                                //1
    std::cin >> n;                                        //1
    std::vector<int> numbers(n);                          //n
    for (int i = 0; i < n; ++i) {                         //2 + 2n
        std::cin >> numbers[i];                           //n
    }

    std::cout << MaxPairwiseProduct(numbers) << "\n";     //1 + 1  + 5 + 4n + [2(n-1)n+1, 2.5 (n-1)n) upper bound
    return 0;                                             // --------------
}                                                         // 11 + 7n + [2(n-1)n+1, 2.5 (n-1)n) upper bound


Writing max_pairwise_product_long.cpp


In [119]:
!g++ max_pairwise_product_long.cpp -o max_pairwise_product_long
!time ./max_pairwise_product_long < input13.txt

120989000

real	0m0.406s
user	0m0.400s
sys	0m0.002s


In [120]:
!g++ max_pairwise_product.cpp -o max_pairwise_product
!time ./max_pairwise_product < input13.txt

120989000

real	0m0.546s
user	0m0.541s
sys	0m0.002s


# Sort solution in C++ and Python

$$ O(n \lg n)$$

**Which sorting algorithm is used in GCC?**
https://stackoverflow.com/questions/7221257/which-sorting-algorithm-is-used-in-gcc

* ...  the algorithm is called introsort, a combination between Insertionsort, Quicksort & Heapsort. ...

**Introsort**
https://en.wikipedia.org/wiki/Introsort

Introsort or introspective sort is a hybrid sorting algorithm that provides both fast average performance and (asymptotically) optimal worst-case performance. It begins with quicksort, it switches to heapsort when the recursion depth exceeds a level based on (the logarithm of) the number of elements being sorted and it switches to insertion sort when the number of elements is below some threshold

*   Class	Sorting algorithm
*   Data structure	Array
*   Worst-case performance	$O(n \lg n)$
*   Average performance	$O(n \lg n)$

**Tinsort**
https://en.wikipedia.org/wiki/Timsort

Timsort is a hybrid, stable sorting algorithm, derived from merge sort and insertion sort, designed to perform well on many kinds of real-world data. It was implemented by Tim Peters in 2002 for use in the Python programming language.

*   Class	Sorting algorithm
*   Data structure	Array
*   Worst-case performance	$O(n \lg n)$
*   Best-case performance $O(n)$
*   Average performance	$O(n \lg n)$
*   Worst-case space complexity $O(n)$

In [110]:
%%writefile max_pairwise_product_with_sort.py
#Python3
if __name__ == "__main__":
    n = int(input())
    lista = list(map(int, input().split()))
    lista.sort()
    print(lista[n - 1] * lista[n - 2])

Writing max_pairwise_product_with_sort.py


In [111]:

%%writefile max_pairwise_product_with_sort_C.cpp
#include
#include
#include

int main() {
    int n;                                                //1
    std::cin >> n;                                        //1
    std::vector numbers(n);                          //n
    for (int i = 0; i < n; ++i) {                         //1 + 2n
        std::cin >> numbers[i];                           //n
    }

    std::sort(numbers.begin(), numbers.end());                          //nlogn
    std::cout << (long long) numbers[n - 1] * numbers[n - 2] << '\n';   //2

    return 0;


}

Writing max_pairwise_product_with_sort_C.cpp


In [121]:
!g++ max_pairwise_product_with_sort_C.cpp -o max_pairwise_product_with_sort_C

max_pairwise_product_with_sort_C.cpp:1:10: error: #include expects "FILENAME" or <FILENAME>
    1 | #include
      |          ^
max_pairwise_product_with_sort_C.cpp:2:10: error: #include expects "FILENAME" or <FILENAME>
    2 | #include
      |          ^
max_pairwise_product_with_sort_C.cpp:3:10: error: #include expects "FILENAME" or <FILENAME>
    3 | #include
      |          ^
max_pairwise_product_with_sort_C.cpp: In function ‘int main()’:
max_pairwise_product_with_sort_C.cpp:7:10: error: ‘cin’ is not a member of ‘std’
    7 |     std::cin >> n;                                        //1
      |          ^~~
max_pairwise_product_with_sort_C.cpp:1:1: note: ‘std::cin’ is defined in header ‘<iostream>’; did you forget to ‘#include <iostream>’?
  +++ |+#include <iostream>
    1 | #include
max_pairwise_product_with_sort_C.cpp:8:10: error: ‘vector’ is not a member of ‘std’
    8 |     std::vector numbers(n);                          //n
      |          ^~~~~~
max_pairwise_product_with_s

Test with inputs 1 and 2



In [124]:
!./max_pairwise_product_with_sort_C < input1.txt
!./max_pairwise_product_with_sort_C < input2.txt

/bin/bash: line 1: ./max_pairwise_product_with_sort_C: No such file or directory
/bin/bash: line 1: ./max_pairwise_product_with_sort_C: No such file or directory


In [123]:
!python max_pairwise_product_with_sort.py < input1.txt
!python max_pairwise_product_with_sort.py < input2.txt

90
90


Test and time it with input 3

In [125]:
!time ./max_pairwise_product_with_sort_C < input3.txt

/bin/bash: line 1: ./max_pairwise_product_with_sort_C: No such file or directory

real	0m0.002s
user	0m0.000s
sys	0m0.001s


In [126]:
%%time
!python max_pairwise_product_with_sort.py < input3.txt

40000000000
CPU times: user 5.21 ms, sys: 1.19 ms, total: 6.4 ms
Wall time: 107 ms


20 stress tests

In [127]:
%%writefile stress_tests.py
#Python3
import random

if __name__ == "__main__":
    for i in range(20):
        n = random.randint(2, 19)
        lista = [random.randint(17, 25631) for i in range(n)]
        lista.sort()
        print(f"n: {n} Product: {lista[n - 1] * lista[n - 2]}")


Writing stress_tests.py


In [128]:
%%writefile stress_tests_C.cpp
#include
#include
#include
#include

int main() {
    std::random_device rd;
    std::mt19937 gen(rd());
    std::uniform_int_distribution n_dist(2, 99);
    std::uniform_int_distribution num_dist(0, 25000);

    for (int i = 0; i < 100; ++i ){
          int n = n_dist(gen);                              //1
          std::vector numbers(n);                          //n
          for (int i = 0; i < n; ++i) {                         //1 + 2n
              numbers[i]= num_dist(gen);                           //n
          }

          std::sort(numbers.begin(), numbers.end());                          //nlogn
          std::cout << "n= "<< n << " Product= " << (long long) numbers[n - 1] * numbers[n - 2] << '\n';   //2
    }
    return 0;
}

Writing stress_tests_C.cpp


In [129]:
!g++ stress_tests_C.cpp -o stress_tests_C
!./stress_tests_C


stress_tests_C.cpp:1:10: error: #include expects "FILENAME" or <FILENAME>
    1 | #include
      |          ^
stress_tests_C.cpp:2:10: error: #include expects "FILENAME" or <FILENAME>
    2 | #include
      |          ^
stress_tests_C.cpp:3:10: error: #include expects "FILENAME" or <FILENAME>
    3 | #include
      |          ^
stress_tests_C.cpp:4:10: error: #include expects "FILENAME" or <FILENAME>
    4 | #include
      |          ^
stress_tests_C.cpp: In function ‘int main()’:
stress_tests_C.cpp:7:10: error: ‘random_device’ is not a member of ‘std’
    7 |     std::random_device rd;
      |          ^~~~~~~~~~~~~
stress_tests_C.cpp:8:10: error: ‘mt19937’ is not a member of ‘std’
    8 |     std::mt19937 gen(rd());
      |          ^~~~~~~
stress_tests_C.cpp:9:10: error: ‘uniform_int_distribution’ is not a member of ‘std’
    9 |     std::uniform_int_distribution n_dist(2, 99);
      |          ^~~~~~~~~~~~~~~~~~~~~~~~
stress_tests_C.cpp:10:10: error: ‘uniform_int_distribution’ is n

In [130]:
!python stress_tests.py

n: 5 Product: 273960385
n: 19 Product: 573170834
n: 12 Product: 625703188
n: 2 Product: 31992212
n: 5 Product: 620049088
n: 18 Product: 556279926
n: 10 Product: 525819097
n: 13 Product: 515115842
n: 2 Product: 20931028
n: 14 Product: 641085032
n: 19 Product: 564749095
n: 12 Product: 573853868
n: 12 Product: 597481612
n: 16 Product: 557696640
n: 12 Product: 448540960
n: 17 Product: 339657948
n: 5 Product: 110894850
n: 19 Product: 531786776
n: 13 Product: 441781893
n: 7 Product: 229599720


Count executed instructions in code and verify with inputs 1,2 and 3


In [131]:
%%writefile max_pairwise_product_with_sort_and_instructions.py
#Python3
import math
if __name__ == "__main__":
    inst=0
    n = int(input())
    inst+=1
    lista = list(map(int, input().split()))
    inst+= 4*n
    lista.sort()
    inst+=  n * math.log2(n)
    print(lista[n - 1] * lista[n - 2])
    inst+=1
    print(f"Instrucciones ejecutadas: {int(inst)}\n")

Writing max_pairwise_product_with_sort_and_instructions.py


In [132]:
%%writefile max_pairwise_product_with_sort_C_and_instructions.cpp
#include
#include
#include
#include

int main() {
    int inst;
    int n;                                                //1
    inst++;
    std::cin >> n;                                        //1
    inst++;
    std::vector numbers(n);                          //n
    inst+=n;
    inst+=1;                                              // add the declaration of i=0
    for (int i = 0; i < n; ++i) {                         //1 + 2n
        inst+=2;
        std::cin >> numbers[i];                           //n
        inst++;
    }

    std::sort(numbers.begin(), numbers.end());                          //nlogn
    inst+=n * log2(n);
    std::cout << (long long) numbers[n - 1] * numbers[n - 2] << '\n';   //2
    inst+=2;
    std::cout << "Instrucciones Ejecutadas: " << inst << '\n';
    return 0;


}

Writing max_pairwise_product_with_sort_C_and_instructions.cpp


In [133]:
!g++ max_pairwise_product_with_sort_C_and_instructions.cpp -o max_pairwise_product_with_sort_C_and_instructions

max_pairwise_product_with_sort_C_and_instructions.cpp:1:10: error: #include expects "FILENAME" or <FILENAME>
    1 | #include
      |          ^
max_pairwise_product_with_sort_C_and_instructions.cpp:2:10: error: #include expects "FILENAME" or <FILENAME>
    2 | #include
      |          ^
max_pairwise_product_with_sort_C_and_instructions.cpp:3:10: error: #include expects "FILENAME" or <FILENAME>
    3 | #include
      |          ^
max_pairwise_product_with_sort_C_and_instructions.cpp:4:10: error: #include expects "FILENAME" or <FILENAME>
    4 | #include
      |          ^
max_pairwise_product_with_sort_C_and_instructions.cpp: In function ‘int main()’:
max_pairwise_product_with_sort_C_and_instructions.cpp:10:10: error: ‘cin’ is not a member of ‘std’
   10 |     std::cin >> n;                                        //1
      |          ^~~
max_pairwise_product_with_sort_C_and_instructions.cpp:1:1: note: ‘std::cin’ is defined in header ‘<iostream>’; did you forget to ‘#include <iostream>

In [134]:
!./max_pairwise_product_with_sort_C_and_instructions < input1.txt
!./max_pairwise_product_with_sort_C_and_instructions < input2.txt
!./max_pairwise_product_with_sort_C_and_instructions < input3.txt

/bin/bash: line 1: ./max_pairwise_product_with_sort_C_and_instructions: No such file or directory
/bin/bash: line 1: ./max_pairwise_product_with_sort_C_and_instructions: No such file or directory
/bin/bash: line 1: ./max_pairwise_product_with_sort_C_and_instructions: No such file or directory


In [135]:
!python max_pairwise_product_with_sort_and_instructions.py < input1.txt
!python max_pairwise_product_with_sort_and_instructions.py < input2.txt
!python max_pairwise_product_with_sort_and_instructions.py < input3.txt

90
Instrucciones ejecutadas: 75

90
Instrucciones ejecutadas: 75

40000000000
Instrucciones ejecutadas: 41



Calculate estimated time by instruction and number of instructions by second

In [136]:
%%writefile max_pairwise_product_with_sort_and_instructions_with_time_and_IbS.py
#Python3
import math
import time
if __name__ == "__main__":
    inicio = time.time()
    inst=0
    n = int(input())
    inst+=1
    lista = list(map(int, input().split()))
    inst+= 4*n
    lista.sort()
    inst+=  n * math.log2(n)
    print(lista[n - 1] * lista[n - 2])
    inst+=1
    print(f"Instrucciones ejecutadas: {int(inst)}")
    fin = time.time()
    tiempo_total = fin - inicio
    instructionTime = tiempo_total/inst
    instructionBySecond= inst/tiempo_total

    print(f"Tiempo total de ejecución: {tiempo_total*1000} milisegundos")
    print(f"Tiempo estimado de cada instruccion: {instructionTime*1000} milisegundos")
    print(f"Instrucciones por segundo: {int(instructionBySecond)} ")

Writing max_pairwise_product_with_sort_and_instructions_with_time_and_IbS.py


In [137]:
!python max_pairwise_product_with_sort_and_instructions_with_time_and_IbS.py < input13.txt

120989000
Instrucciones ejecutadas: 191679
Tiempo total de ejecución: 2.5141239166259766 milisegundos
Tiempo estimado de cada instruccion: 1.3116298597392513e-05 milisegundos
Instrucciones por segundo: 76241021 


### Estimate and verify the maximun n that can be solved in 1 second

In [138]:
%%writefile maximun_n_that_can_be_solved.py
#Python3
import random

if __name__ == "__main__":
      n = 600000
      lista = [random.randint(17, 25631) for i in range(n)]
      lista.sort()
      print(f"n: {n} Product: {lista[n - 1] * lista[n - 2]}")

Writing maximun_n_that_can_be_solved.py


In [139]:
%%time
!python maximun_n_that_can_be_solved.py

n: 600000 Product: 656948161
CPU times: user 13.4 ms, sys: 0 ns, total: 13.4 ms
Wall time: 908 ms


In [140]:
%%writefile maximun_n_that_can_be_solved_C.cpp
#include
#include
#include
#include

int main() {
    std::random_device rd;
    std::mt19937 gen(rd());
    std::uniform_int_distribution n_dist(2, 99);
    std::uniform_int_distribution num_dist(0, 25000);
    int n = 1700000;                              //1
    std::vector numbers(n);                          //n
    for (int i = 0; i < n; ++i) {                         //1 + 2n
      numbers[i]= num_dist(gen);                           //n
    }

    std::sort(numbers.begin(), numbers.end());                          //nlogn
    std::cout << "n= "<< n << " Product= " << (long long) numbers[n - 1] * numbers[n - 2] << '\n';   //2
    return 0;

}

Writing maximun_n_that_can_be_solved_C.cpp


In [141]:
!g++ maximun_n_that_can_be_solved_C.cpp -o maximun_n_that_can_be_solved_C
!time ./maximun_n_that_can_be_solved_C

maximun_n_that_can_be_solved_C.cpp:1:10: error: #include expects "FILENAME" or <FILENAME>
    1 | #include
      |          ^
maximun_n_that_can_be_solved_C.cpp:2:10: error: #include expects "FILENAME" or <FILENAME>
    2 | #include
      |          ^
maximun_n_that_can_be_solved_C.cpp:3:10: error: #include expects "FILENAME" or <FILENAME>
    3 | #include
      |          ^
maximun_n_that_can_be_solved_C.cpp:4:10: error: #include expects "FILENAME" or <FILENAME>
    4 | #include
      |          ^
maximun_n_that_can_be_solved_C.cpp: In function ‘int main()’:
maximun_n_that_can_be_solved_C.cpp:7:10: error: ‘random_device’ is not a member of ‘std’
    7 |     std::random_device rd;
      |          ^~~~~~~~~~~~~
maximun_n_that_can_be_solved_C.cpp:8:10: error: ‘mt19937’ is not a member of ‘std’
    8 |     std::mt19937 gen(rd());
      |          ^~~~~~~
maximun_n_that_can_be_solved_C.cpp:9:10: error: ‘uniform_int_distribution’ is not a member of ‘std’
    9 |     std::uniform_int_distri

## Efficient solution with largest and second-largest in an list, array or vector in C++ and Python

 $$T(n) < 10 n  = O(n)$$





Add in each line a coment with the number of intstrcution that are executed in that line

In [142]:
%%writefile max_pairwise_product_long_efficient.cpp
#include
#include
#include

long long MaxPairwiseProduct(const std::vector& numbers) {
    long long max_product = 0LL;                                             //1
	  long long n_major = 0LL;
    long long n_major_2 = 0LL;                                             //1
    int n = numbers.size();                                                  //1

    for(int i = 0; i < n; ++i) {                   //1 + 2n
      if (numbers[i] > n_major){
        n_major_2 = n_major ;                //n
        n_major = numbers[i];
      }
      else if (numbers[i] > n_major_2){                   //n
        n_major_2 = numbers[i];                    //[0, n]

      }
    }
    max_product=n_major*n_major_2;                                                                          // --------------
    return max_product;
}

int main() {
    int n;                                                //1
    std::cin >> n;                                        //1
    std::vector numbers(n);                          //n
    for (int i = 0; i < n; ++i) {                         //1 + 2n
        std::cin >> numbers[i];                           //n
    }

    std::cout << MaxPairwiseProduct(numbers) << "\n";
    return 0;
}

Writing max_pairwise_product_long_efficient.cpp


In [143]:
!g++ max_pairwise_product_long_efficient.cpp -o max_pairwise_product_long_efficient

max_pairwise_product_long_efficient.cpp:1:10: error: #include expects "FILENAME" or <FILENAME>
    1 | #include
      |          ^
max_pairwise_product_long_efficient.cpp:2:10: error: #include expects "FILENAME" or <FILENAME>
    2 | #include
      |          ^
max_pairwise_product_long_efficient.cpp:3:10: error: #include expects "FILENAME" or <FILENAME>
    3 | #include
      |          ^
max_pairwise_product_long_efficient.cpp:5:41: error: ‘vector’ in namespace ‘std’ does not name a type
    5 | long long MaxPairwiseProduct(const std::vector& numbers) {
      |                                         ^~~~~~
max_pairwise_product_long_efficient.cpp:1:1: note: ‘std::vector’ is defined in header ‘<vector>’; did you forget to ‘#include <vector>’?
  +++ |+#include <vector>
    1 | #include
max_pairwise_product_long_efficient.cpp: In function ‘long long int MaxPairwiseProduct(const int&)’:
max_pairwise_product_long_efficient.cpp:9:21: error: request for member ‘size’ in ‘numbers’, which is 

In [144]:
%%writefile max_pairwise_product_efficient.py
#Python3
def max_pairwise_product(lista):
    n_major = 0
    n_major_2 = 0
    max_product = 0
    for i in lista:
        if i > n_major:
            n_major_2=n_major
            n_major = i
        elif i > n_major_2:
            n_major_2=i
    max_product = n_major_2 * n_major

    return max_product

if __name__ == "__main__":
    n = int(input())
    lista = list(map(int, input().split()))
    print(max_pairwise_product(lista))

Writing max_pairwise_product_efficient.py


Test with inputs 1 and 2



In [145]:
!./max_pairwise_product_long_efficient < input1.txt
!./max_pairwise_product_long_efficient < input2.txt

/bin/bash: line 1: ./max_pairwise_product_long_efficient: No such file or directory
/bin/bash: line 1: ./max_pairwise_product_long_efficient: No such file or directory


In [147]:
!python max_pairwise_product_efficient.py < input1.txt
!python max_pairwise_product_efficient.py < input2.txt

90
90


Test and time it with input 3

In [148]:
%%time
!python max_pairwise_product_efficient.py < input3.txt

40000000000
CPU times: user 4.63 ms, sys: 1.9 ms, total: 6.52 ms
Wall time: 107 ms


In [149]:
!time ./max_pairwise_product_long_efficient < input3.txt

/bin/bash: line 1: ./max_pairwise_product_long_efficient: No such file or directory

real	0m0.001s
user	0m0.001s
sys	0m0.000s


20 stress tests

In [150]:
%%writefile max_pairwise_product_long_efficient_stress_test.cpp
#include
#include
#include
#include

long long MaxPairwiseProduct(const std::vector& numbers) {
    long long max_product = 0LL;                                             //1
	  long long n_major = 0LL;
    long long n_major_2 = 0LL;                                             //1
    int n = numbers.size();                                                  //1

    for(int i = 0; i < n; ++i) {                   //1 + 2n
      if (numbers[i] > n_major){
        n_major_2 = n_major ;                //n
        n_major = numbers[i];
      }
      else if (numbers[i] > n_major_2){                   //n
        n_major_2 = numbers[i];                    //[0, n]

      }
    }
    max_product=n_major*n_major_2;                                                                          // --------------
    return max_product;
}

int main() {
    std::random_device rd;
    std::mt19937 gen(rd());
    std::uniform_int_distribution n_dist(2, 19);
    std::uniform_int_distribution num_dist(0, 25000);
    int n;                                                //1
    std::vector numbers(n);                          //n
    for (int i = 0; i < 20; ++i ){
      int n = n_dist(gen);                              //1
      std::vector numbers(n);                          //n
      for (int i = 0; i < n; ++i) {                         //1 + 2n
        numbers[i]= num_dist(gen);                           //n
      }
      std::cout << "n= "<< n << " Product= " << MaxPairwiseProduct(numbers) << "\n";
    }
    return 0;
}

Writing max_pairwise_product_long_efficient_stress_test.cpp


In [151]:
!g++ max_pairwise_product_long_efficient_stress_test.cpp -o max_pairwise_product_long_efficient_stress_test
!./max_pairwise_product_long_efficient_stress_test

max_pairwise_product_long_efficient_stress_test.cpp:1:10: error: #include expects "FILENAME" or <FILENAME>
    1 | #include
      |          ^
max_pairwise_product_long_efficient_stress_test.cpp:2:10: error: #include expects "FILENAME" or <FILENAME>
    2 | #include
      |          ^
max_pairwise_product_long_efficient_stress_test.cpp:3:10: error: #include expects "FILENAME" or <FILENAME>
    3 | #include
      |          ^
max_pairwise_product_long_efficient_stress_test.cpp:4:10: error: #include expects "FILENAME" or <FILENAME>
    4 | #include
      |          ^
max_pairwise_product_long_efficient_stress_test.cpp:6:41: error: ‘vector’ in namespace ‘std’ does not name a type
    6 | long long MaxPairwiseProduct(const std::vector& numbers) {
      |                                         ^~~~~~
max_pairwise_product_long_efficient_stress_test.cpp:1:1: note: ‘std::vector’ is defined in header ‘<vector>’; did you forget to ‘#include <vector>’?
  +++ |+#include <vector>
    1 | #include


In [153]:
%%writefile max_pairwise_product_efficient_stress_tests.py
#Python3
import random
def max_pairwise_product(lista):
    n_major = 0
    n_major_2 = 0
    max_product = 0
    for i in lista:
        if i > n_major:
            n_major_2=n_major
            n_major = i
        elif i > n_major_2:
            n_major_2=i
    max_product = n_major_2 * n_major

    return max_product

if __name__ == "__main__":
    for i in range(20):
      n = random.randint(2, 19)
      lista = [random.randint(17, 25631) for i in range(n)]
      print(f"n=  {n}   Product= {max_pairwise_product(lista)}")

Overwriting max_pairwise_product_efficient_stress_tests.py


In [154]:
!python max_pairwise_product_efficient_stress_tests.py

n=  18   Product= 431322144
n=  4   Product= 534855104
n=  15   Product= 472352348
n=  11   Product= 509294610
n=  4   Product= 475166000
n=  7   Product= 368063955
n=  14   Product= 504208656
n=  10   Product= 458005436
n=  15   Product= 616776620
n=  11   Product= 580309757
n=  13   Product= 596517194
n=  16   Product= 357885428
n=  2   Product= 30563669
n=  8   Product= 364508276
n=  12   Product= 649049190
n=  6   Product= 467920245
n=  15   Product= 633305652
n=  16   Product= 556935300
n=  5   Product= 108895840
n=  12   Product= 343762265


Count executed instructions in code and verify with inputs 1,2 and 3


In [155]:
%%writefile max_pairwise_product_long_efficient_and_instructions.cpp
#include
#include
#include

long long inst= 0LL;

long long MaxPairwiseProduct(const std::vector& numbers) {
    long long max_product = 0LL;                                             //1
    inst++;
	  long long n_major = 0LL;
    inst++;
    long long n_major_2 = 0LL;                                             //1
    inst++;
    int n = numbers.size();                                                  //1
    inst++;
    inst++;
    for(int i = 0; i < n; ++i) {                   //1 + 2n
      inst+=2;
      inst++;
      if (numbers[i] > n_major){
        inst++;
        n_major_2 = n_major ;                //n
        inst++;
        n_major = numbers[i];
        inst++;
      }
      else if (numbers[i] > n_major_2){
        inst++;
        n_major_2 = numbers[i];                    //[0, n]
        inst++;
        inst++;
      }
    }
    max_product=n_major*n_major_2;
    inst++;                                                                         // --------------
    return max_product;
}
int main() {
    int n;                                                //1
    inst++;
    std::cin >> n;                                        //1
    inst++;
    std::vector numbers(n);                          //n
    inst++;
    inst++;
    for (int i = 0; i < n; ++i) {                         //1 + 2n
        std::cin >> numbers[i];                           //n
    }
    inst+=n;

    std::cout << MaxPairwiseProduct(numbers) << "\n";
    inst++;
    std::cout <<"Numero de instrucciones: "<< inst << "\n";
    return 0;
}

Writing max_pairwise_product_long_efficient_and_instructions.cpp


In [156]:
!g++ max_pairwise_product_long_efficient_and_instructions.cpp -o max_pairwise_product_long_efficient_and_instructions

max_pairwise_product_long_efficient_and_instructions.cpp:1:10: error: #include expects "FILENAME" or <FILENAME>
    1 | #include
      |          ^
max_pairwise_product_long_efficient_and_instructions.cpp:2:10: error: #include expects "FILENAME" or <FILENAME>
    2 | #include
      |          ^
max_pairwise_product_long_efficient_and_instructions.cpp:3:10: error: #include expects "FILENAME" or <FILENAME>
    3 | #include
      |          ^
max_pairwise_product_long_efficient_and_instructions.cpp:7:41: error: ‘vector’ in namespace ‘std’ does not name a type
    7 | long long MaxPairwiseProduct(const std::vector& numbers) {
      |                                         ^~~~~~
max_pairwise_product_long_efficient_and_instructions.cpp:1:1: note: ‘std::vector’ is defined in header ‘<vector>’; did you forget to ‘#include <vector>’?
  +++ |+#include <vector>
    1 | #include
max_pairwise_product_long_efficient_and_instructions.cpp: In function ‘long long int MaxPairwiseProduct(const int&)’:


In [157]:
!./max_pairwise_product_long_efficient_and_instructions < input1.txt
!./max_pairwise_product_long_efficient_and_instructions < input2.txt
!./max_pairwise_product_long_efficient_and_instructions < input3.txt

/bin/bash: line 1: ./max_pairwise_product_long_efficient_and_instructions: No such file or directory
/bin/bash: line 1: ./max_pairwise_product_long_efficient_and_instructions: No such file or directory
/bin/bash: line 1: ./max_pairwise_product_long_efficient_and_instructions: No such file or directory


In [158]:
%%writefile max_pairwise_product_efficient_and_instructions.py
#Python3

inst=0
def max_pairwise_product(lista):
    global inst
    n_major = 0
    inst+=1
    n_major_2 = 0
    inst+=1
    max_product = 0
    inst+=1
    for i in lista:
        inst+=1
        if i > n_major:
            inst+=1
            n_major_2=n_major
            inst+=1
            n_major = i
            inst+=1
        elif i > n_major_2:
            inst+=1
            n_major_2=i
            inst+=1
            inst+=1
    max_product = n_major_2 * n_major
    inst+=1

    return max_product

if __name__ == "__main__":
    n = int(input())
    inst+=1
    lista = list(map(int, input().split()))
    inst+=4*n
    print(max_pairwise_product(lista))
    inst+=1
    print(f"Numero Total de Instrucciones:  {inst}")


Writing max_pairwise_product_efficient_and_instructions.py


In [159]:
!python max_pairwise_product_efficient_and_instructions.py < input1.txt
!python max_pairwise_product_efficient_and_instructions.py < input2.txt
!python max_pairwise_product_efficient_and_instructions.py < input3.txt

90
Numero Total de Instrucciones:  86
90
Numero Total de Instrucciones:  62
40000000000
Numero Total de Instrucciones:  51


Calculate estimated time by instruction and number of instructions by second

In [160]:
%%writefile max_pairwise_product_long_efficient_and_instructions_by_second.cpp
#include
#include
#include
#include

long long inst= 0LL;

long long MaxPairwiseProduct(const std::vector& numbers) {
    long long max_product = 0LL;                                             //1
    inst++;
	  long long n_major = 0LL;
    inst++;
    long long n_major_2 = 0LL;                                             //1
    inst++;
    int n = numbers.size();                                                  //1
    inst++;
    inst++;
    for(int i = 0; i < n; ++i) {                   //1 + 2n
      inst+=2;
      inst++;
      if (numbers[i] > n_major){
        inst++;
        n_major_2 = n_major ;                //n
        inst++;
        n_major = numbers[i];
        inst++;
      }
      else if (numbers[i] > n_major_2){
        inst++;
        n_major_2 = numbers[i];                    //[0, n]
        inst++;
        inst++;
      }
    }
    max_product=n_major*n_major_2;
    inst++;                                                                         // --------------
    return max_product;
}
int main() {
    auto start_time = std::chrono::high_resolution_clock::now();
    int n;                                                //1
    inst++;
    std::cin >> n;                                        //1
    inst++;
    std::vector numbers(n);                          //n
    inst++;
    inst++;
    for (int i = 0; i < n; ++i) {                         //1 + 2n
        std::cin >> numbers[i];                           //n
    }
    inst+=n;

    std::cout << MaxPairwiseProduct(numbers) << "\n";
    inst++;
    auto end_time = std::chrono::high_resolution_clock::now();
    auto duration = std::chrono::duration_cast(end_time - start_time);
    float duration_sec,duration_f;
    duration_f=static_cast(duration.count());
    duration_sec = duration_f/1000;
    std::cout <<"Numero de instrucciones: "<< inst << "\n";
    std::cout <<"Duracion Estiamda: "<< duration_sec << " segundos" <<"\n";
    std::cout <<"Duracion Estiamda por instruccion: "<< duration_sec/inst << " segundos" <<"\n";
    std::cout <<"Numero de instrucciones por segundo: "<< 1/(duration_sec/inst) << " Instrucciones" <<"\n";
    return 0;
}

Writing max_pairwise_product_long_efficient_and_instructions_by_second.cpp


In [161]:
!g++ max_pairwise_product_long_efficient_and_instructions_by_second.cpp -o max_pairwise_product_long_efficient_and_instructions_by_second


max_pairwise_product_long_efficient_and_instructions_by_second.cpp:1:10: error: #include expects "FILENAME" or <FILENAME>
    1 | #include
      |          ^
max_pairwise_product_long_efficient_and_instructions_by_second.cpp:2:10: error: #include expects "FILENAME" or <FILENAME>
    2 | #include
      |          ^
max_pairwise_product_long_efficient_and_instructions_by_second.cpp:3:10: error: #include expects "FILENAME" or <FILENAME>
    3 | #include
      |          ^
max_pairwise_product_long_efficient_and_instructions_by_second.cpp:4:10: error: #include expects "FILENAME" or <FILENAME>
    4 | #include
      |          ^
max_pairwise_product_long_efficient_and_instructions_by_second.cpp:8:41: error: ‘vector’ in namespace ‘std’ does not name a type
    8 | long long MaxPairwiseProduct(const std::vector& numbers) {
      |                                         ^~~~~~
max_pairwise_product_long_efficient_and_instructions_by_second.cpp:1:1: note: ‘std::vector’ is defined in header ‘<ve

In [162]:
!./max_pairwise_product_long_efficient_and_instructions_by_second < input13.txt

/bin/bash: line 1: ./max_pairwise_product_long_efficient_and_instructions_by_second: No such file or directory


In [164]:
%%writefile max_pairwise_product_efficient_and_instructions_by_second.py
#Python3
import time

inst=0
def max_pairwise_product(lista):
    global inst
    n_major = 0
    inst+=1
    n_major_2 = 0
    inst+=1
    max_product = 0
    inst+=1
    for i in lista:
        inst+=1
        if i > n_major:
            inst+=1
            n_major_2=n_major
            inst+=1
            n_major = i
            inst+=1
        elif i > n_major_2:
            inst+=1
            n_major_2=i
            inst+=1
            inst+=1
    max_product = n_major_2 * n_major
    inst+=1

    return max_product

if __name__ == "__main__":
    inicio=time.time()
    n = int(input())
    inst+=1
    lista = list(map(int, input().split()))
    inst+=4*n
    print(max_pairwise_product(lista))
    inst+=1
    fin = time.time()
    tiempo_total = fin - inicio
    instructionTime = tiempo_total/inst
    instructionBySecond= 1/instructionTime
    print(f"Numero Total de Instrucciones:  {inst}")
    print(f"Tiempo total de ejecución: {tiempo_total} segundos")
    print(f"Tiempo estimado de cada instruccion: {instructionTime} segundos")
    print(f"Instrucciones por segundo: {int(instructionBySecond)} ")


Overwriting max_pairwise_product_efficient_and_instructions_by_second.py


In [165]:
!python max_pairwise_product_efficient_and_instructions_by_second.py < input13.txt

120989000
Numero Total de Instrucciones:  88006
Tiempo total de ejecución: 0.005986213684082031 segundos
Tiempo estimado de cada instruccion: 6.80205177383591e-08 segundos
Instrucciones por segundo: 14701446 


### Estimate and verify the maximun n that can be solved in 1 second

In [166]:
%%writefile maximun_n_that_can_be_solved_efficient.py
#Python3
import random
def max_pairwise_product(lista):
    n_major = 0
    n_major_2 = 0
    max_product = 0
    for i in lista:
        if i > n_major:
            n_major_2=n_major
            n_major = i
        elif i > n_major_2:
            n_major_2=i
    max_product = n_major_2 * n_major

    return max_product

if __name__ == "__main__":
    n = 800000
    lista = [random.randint(17, 25631) for i in range(n)]
    print(f"n=  {n}   Product= {max_pairwise_product(lista)}")

Writing maximun_n_that_can_be_solved_efficient.py


In [167]:
%%time
!python maximun_n_that_can_be_solved_efficient.py

n=  800000   Product= 656948161
CPU times: user 12.3 ms, sys: 2.99 ms, total: 15.3 ms
Wall time: 1.01 s


## Super efficient solution

Whitout storing the list in memory, only stpring the largest and second largest and whitout calling a function, add in each line a coment with the number of instructions that are executed in that line

in C++ and Python


 $$T(n) < 13 n  = O(n)$$


In [107]:
%%writefile max_pairwise_super_efficient.cpp
#include
#include
#include
using namespace std;

int main() {
    int n;                                     // 1
    cin >> n;                                  // 1
    int numero_M = 0;                   // 1
    int segundo_Numero_M = 0;             // 1


    for (int i = 0; i < n; ++i) {              //1 + 2n
        int comp = 0;                           // n
        cin >> comp;                            // n

        if (comp > numero_M){                  // n
          segundo_Numero_M = numero_M;         //n
          numero_M = comp;                     //n

        } else if (comp > segundo_Numero_M) {    // n
          segundo_Numero_M = comp;             //n
        }

    }

    cout  << numero_M * segundo_Numero_M << "\n"; // 1
    return 0;
}

Writing max_pairwise_super_efficient.cpp


In [108]:
!g++ max_pairwise_super_efficient.cpp -o max_pairwise_super_efficient

max_pairwise_super_efficient.cpp:1:10: error: #include expects "FILENAME" or <FILENAME>
    1 | #include
      |          ^
max_pairwise_super_efficient.cpp:2:10: error: #include expects "FILENAME" or <FILENAME>
    2 | #include
      |          ^
max_pairwise_super_efficient.cpp:3:10: error: #include expects "FILENAME" or <FILENAME>
    3 | #include
      |          ^
max_pairwise_super_efficient.cpp: In function ‘int main()’:
max_pairwise_super_efficient.cpp:8:5: error: ‘cin’ was not declared in this scope
    8 |     cin >> n;                                  // 1
      |     ^~~
max_pairwise_super_efficient.cpp:1:1: note: ‘std::cin’ is defined in header ‘<iostream>’; did you forget to ‘#include <iostream>’?
  +++ |+#include <iostream>
    1 | #include
max_pairwise_super_efficient.cpp:27:5: error: ‘cout’ was not declared in this scope
   27 |     cout  << numero_M * segundo_Numero_M << "\n"; // 1
      |     ^~~~
max_pairwise_super_efficient.cpp:27:5: note: ‘std::cout’ is defined in

Test with inputs 1 and 2




In [109]:
!./max_pairwise_super_efficient < input1.txt
!./max_pairwise_super_efficient < input2.txt

/bin/bash: line 1: ./max_pairwise_super_efficient: No such file or directory
/bin/bash: line 1: ./max_pairwise_super_efficient: No such file or directory


Test and time it with input 3

In [168]:
!time ./max_pairwise_super_efficient < input3.txt

/bin/bash: line 1: ./max_pairwise_super_efficient: No such file or directory

real	0m0.001s
user	0m0.000s
sys	0m0.001s


20 stress tests

In [169]:
%%writefile max_pairwise_super_efficient_stress_tests.cpp
#include
#include
#include
#include
using namespace std;

int main() {
    std::random_device rd;
    std::mt19937 gen(rd());
    std::uniform_int_distribution n_dist(2, 19);
    std::uniform_int_distribution num_dist(0, 25000);
    int n;                                     // 1



    for (int i = 0; i < 20; ++i ){

          int n = n_dist(gen);
          int numero_M = 0;
          int segundo_Numero_M = 0;
          for (int i = 0; i < n; ++i) {
            int comp = 0;                           // n
            comp= num_dist(gen);                    //1 + 2n
            if (comp > numero_M){                  // n
              segundo_Numero_M = numero_M;         //n
              numero_M = comp;                     //n

            } else if (comp > segundo_Numero_M) {    // n
              segundo_Numero_M = comp;             //n
            }


          }

        cout << "n= "<< n << " Multiplicacion= " << numero_M * segundo_Numero_M  << "\n";
    }
    return 0;
}

Writing max_pairwise_super_efficient_stress_tests.cpp


In [170]:
!g++ max_pairwise_super_efficient_stress_tests.cpp -o max_pairwise_super_efficient_stress_tests
!./max_pairwise_super_efficient_stress_tests

max_pairwise_super_efficient_stress_tests.cpp:1:10: error: #include expects "FILENAME" or <FILENAME>
    1 | #include
      |          ^
max_pairwise_super_efficient_stress_tests.cpp:2:10: error: #include expects "FILENAME" or <FILENAME>
    2 | #include
      |          ^
max_pairwise_super_efficient_stress_tests.cpp:3:10: error: #include expects "FILENAME" or <FILENAME>
    3 | #include
      |          ^
max_pairwise_super_efficient_stress_tests.cpp:4:10: error: #include expects "FILENAME" or <FILENAME>
    4 | #include
      |          ^
max_pairwise_super_efficient_stress_tests.cpp: In function ‘int main()’:
max_pairwise_super_efficient_stress_tests.cpp:8:10: error: ‘random_device’ is not a member of ‘std’
    8 |     std::random_device rd;
      |          ^~~~~~~~~~~~~
max_pairwise_super_efficient_stress_tests.cpp:9:10: error: ‘mt19937’ is not a member of ‘std’
    9 |     std::mt19937 gen(rd());
      |          ^~~~~~~
max_pairwise_super_efficient_stress_tests.cpp:10:10: error

Count executed instructions in code and verify with inputs 1,2 and 3

In [171]:
%%writefile max_pairwise_super_efficient_count.cpp
#include
#include
#include
using namespace std;

int main() {
    int inst = 0;
    int n;                                     // 1
    inst++;
    cin >> n;                                  // 1
    inst++;
    int numero_M = 0;                   // 1
    inst++;
    int segundo_Numero_M = 0;             // 1
    inst++;

    inst++;
    inst+=2*n;
    for (int i = 0; i < n; ++i) {              //1 + 2n
        int comp = 0;                           // n
        inst++;
        cin >> comp;                            // n
        inst++;
        inst++;
        if (comp > numero_M){                  // n
          segundo_Numero_M = numero_M;         //0 o n
          inst++;
          numero_M = comp;                     //0 o n
          inst++;

        } else if (comp > segundo_Numero_M) {    // 0 o n
          inst++;
          segundo_Numero_M = comp;             // 0 o n
          inst++;
        }
    }

    cout  << numero_M * segundo_Numero_M << "\n"; // 1
    inst++;
    cout << "Numero de Instrucciones: " << inst << "\n";
    return 0;
}

Writing max_pairwise_super_efficient_count.cpp


In [172]:
!g++ max_pairwise_super_efficient_count.cpp -o max_pairwise_super_efficient_count

max_pairwise_super_efficient_count.cpp:1:10: error: #include expects "FILENAME" or <FILENAME>
    1 | #include
      |          ^
max_pairwise_super_efficient_count.cpp:2:10: error: #include expects "FILENAME" or <FILENAME>
    2 | #include
      |          ^
max_pairwise_super_efficient_count.cpp:3:10: error: #include expects "FILENAME" or <FILENAME>
    3 | #include
      |          ^
max_pairwise_super_efficient_count.cpp: In function ‘int main()’:
max_pairwise_super_efficient_count.cpp:10:5: error: ‘cin’ was not declared in this scope
   10 |     cin >> n;                                  // 1
      |     ^~~
max_pairwise_super_efficient_count.cpp:1:1: note: ‘std::cin’ is defined in header ‘<iostream>’; did you forget to ‘#include <iostream>’?
  +++ |+#include <iostream>
    1 | #include
max_pairwise_super_efficient_count.cpp:38:5: error: ‘cout’ was not declared in this scope
   38 |     cout  << numero_M * segundo_Numero_M << "\n"; // 1
      |     ^~~~
max_pairwise_super_efficien

In [173]:
!./max_pairwise_super_efficient_count < input1.txt
!./max_pairwise_super_efficient_count < input2.txt
!./max_pairwise_super_efficient_count < input3.txt

/bin/bash: line 1: ./max_pairwise_super_efficient_count: No such file or directory
/bin/bash: line 1: ./max_pairwise_super_efficient_count: No such file or directory
/bin/bash: line 1: ./max_pairwise_super_efficient_count: No such file or directory


Calculate estimated time by instruction and number of instructions by second

In [174]:
t_por_instruccion = 0.003 / 46
instrucciones_por_segundo = 1 / t_por_instruccion
print(f"Tiempo por instruccion: {t_por_instruccion}")
print(f"Instrucciones por segundo: {instrucciones_por_segundo}")

Tiempo por instruccion: 6.521739130434783e-05
Instrucciones por segundo: 15333.333333333332


### Super efficient solution in Python

Whitout storing the list in memory and whitout calling a function, add in each line a coment with the number of intstrcution that are executed in that line

In [94]:

%%writefile max_pairwise_super_efficient.py
n = int(input())                            # 1
numero_M = 0                                # 1
segundo_Numero_M = 0                        # 1

for comp in input().split():                 # 3 + 2n

    comp = int(comp)                          # n

    if comp > numero_M:                    # n
        segundo_Numero_M = numero_M      # 0 o n
        numero_M = comp                  # 0 o n
    elif comp > segundo_Numero_M:           # n
        segundo_Numero_M = comp             # 0 o n
print(f"{numero_M * segundo_Numero_M}")

Overwriting max_pairwise_super_efficient.py


Test with inputs 1 and 2




In [95]:
!python max_pairwise_super_efficient.py < input1.txt
!python max_pairwise_super_efficient.py < input2.txt

90
90


Test and time it with input 3

In [100]:
%%time
!python max_pairwise_super_efficient.py < input3.txt

40000000000
CPU times: user 6.19 ms, sys: 0 ns, total: 6.19 ms
Wall time: 106 ms


20 stress tests

In [101]:
%%writefile max_pairwise_super_efficient_stress_tests.py
import random

for i in range(20):
  n = random.randint(2, 19)
  numero_M = 0                                # 1
  segundo_Numero_M = 0
  for j in range(n):
    comp=random.randint(17, 25631)
    if comp > numero_M:                    # n
      segundo_Numero_M = numero_M      # 0 o n
      numero_M = comp                  # 0 o n
    elif comp > segundo_Numero_M:           # n
      segundo_Numero_M = comp             # 0 o n
  print(f"{i+1}. n= {n} Producto= {numero_M * segundo_Numero_M}")

Writing max_pairwise_super_efficient_stress_tests.py


In [102]:
!python max_pairwise_super_efficient_stress_tests.py

1. n= 7 Producto= 386472988
2. n= 16 Producto= 503578816
3. n= 5 Producto= 547337813
4. n= 19 Producto= 542860875
5. n= 19 Producto= 465455344
6. n= 9 Producto= 388107276
7. n= 10 Producto= 599805550
8. n= 14 Producto= 514385248
9. n= 16 Producto= 550112780
10. n= 6 Producto= 568760038
11. n= 9 Producto= 458977424
12. n= 14 Producto= 569582235
13. n= 4 Producto= 593145630
14. n= 14 Producto= 578246449
15. n= 15 Producto= 437926176
16. n= 2 Producto= 345192461
17. n= 19 Producto= 498801310
18. n= 12 Producto= 575204580
19. n= 6 Producto= 458151512
20. n= 16 Producto= 412795350


Count executed instructions in code and verify with inputs 1,2 and 3

In [105]:
%%writefile max_pairwise_super_efficient_count.py
instrucciones=0
n = int(input())                            # 1
instrucciones+=1
numero_M = 0                                # 1
instrucciones+=1
segundo_Numero_M = 0                        # 1
instrucciones+=1
instrucciones+=1
instrucciones+=2*n
for comp in input().split():                 # 1 + 2n

    comp = int(comp)                          # n
    instrucciones+=1

    instrucciones+=1
    if comp > numero_M:                    # n
        segundo_Numero_M = numero_M      # 0 o n
        instrucciones+=1
        numero_M = comp                  # 0 o n
        instrucciones+=1
    elif comp > segundo_Numero_M:           # 0 o n
        instrucciones+=1
        segundo_Numero_M = comp             # 0 o n
        instrucciones+=1

print(f"{numero_M * segundo_Numero_M}")
print(f"Instrucciones totales: {instrucciones} ")

Overwriting max_pairwise_super_efficient_count.py


In [175]:
!python max_pairwise_super_efficient_count.py < input1.txt
!python max_pairwise_super_efficient_count.py < input2.txt
!python max_pairwise_super_efficient_count.py < input3.txt

90
Instrucciones totales: 64 
90
Instrucciones totales: 48 
40000000000
Instrucciones totales: 38 


## Calculate estimated time by instruction and number of instructions by second

In [106]:
t_por_instruccion = 0.00405 / 38
instrucciones_por_segundo = 1 / t_por_instruccion
print(f"Tiempo por instruccion: {t_por_instruccion}")
print(f"Instrucciones por segundo: {instrucciones_por_segundo}")

Tiempo por instruccion: 0.00010657894736842105
Instrucciones por segundo: 9382.716049382716


### Estimate and verify the maximun n that can be solved in 1 second

In [96]:

%%writefile maximun_n_that_can_be_solved_super_efficient.py
import random

n = 900000
numero_M = 0                                # 1
segundo_Numero_M = 0
for j in range(n):
  comp=random.randint(17, 25631)
  if comp > numero_M:                    # n
    segundo_Numero_M = numero_M      # 0 o n
    numero_M = comp                  # 0 o n
  elif comp > segundo_Numero_M:           # n
    segundo_Numero_M = comp             # 0 o n
print(f"Producto= {numero_M * segundo_Numero_M}")

Overwriting maximun_n_that_can_be_solved_super_efficient.py


In [97]:
%%time
!python maximun_n_that_can_be_solved_super_efficient.py

Producto= 656948161
CPU times: user 15.9 ms, sys: 2.56 ms, total: 18.4 ms
Wall time: 1.21 s


## Perform 100 randomized automated tests in C++ and Python with a porgram  compares the answers,for the same 100 random input instances, of the super efficient solution $(T(n) < 10 n = O(n))$ against the sorting solution  $(T(n)  = O(n \lg{n}))$,  if the answers are not the same stops and displays "Error" listing  the input and the two answers; if the 100 tests are completed without stoping displays "Pass".

In [98]:
%%writefile max_pairwise_super_efficient_against_sorting_stress_tests.py
import random

for i in range(100):
  n = random.randint(2, 99)
  numero_M = 0                                # 1
  segundo_Numero_M = 0
  lista = []
  for j in range(n):
    comp=random.randint(17, 25631)
    lista.append(comp)
    if comp > numero_M:                    # n
      segundo_Numero_M = numero_M      # 0 o n
      numero_M = comp                  # 0 o n
    elif comp > segundo_Numero_M:           # n
      segundo_Numero_M = comp             # 0 o n
  lista.sort()
  sol_sorting=lista[n - 1] * lista[n - 2]
  sol_super_e = numero_M * segundo_Numero_M


  if sol_sorting != sol_super_e:
    print("Error")
    break
  else:
      print(f"{i+1}. SUPER EFFICENT SOLITION: n= {n} Product= {numero_M * segundo_Numero_M}")
      print(f"{i+1}. SORTING SOLUTION: n= {n} Product: {lista[n - 1] * lista[n - 2]}")
  if i==99:
    print("Pass")

Overwriting max_pairwise_super_efficient_against_sorting_stress_tests.py


In [99]:
!python max_pairwise_super_efficient_against_sorting_stress_tests.py

1. SUPER EFFICENT SOLITION: n= 35 Product= 581987830
1. SORTING SOLUTION: n= 35 Product: 581987830
2. SUPER EFFICENT SOLITION: n= 2 Product= 201390371
2. SORTING SOLUTION: n= 2 Product: 201390371
3. SUPER EFFICENT SOLITION: n= 26 Product= 556652580
3. SORTING SOLUTION: n= 26 Product: 556652580
4. SUPER EFFICENT SOLITION: n= 63 Product= 625542200
4. SORTING SOLUTION: n= 63 Product: 625542200
5. SUPER EFFICENT SOLITION: n= 21 Product= 591906000
5. SORTING SOLUTION: n= 21 Product: 591906000
6. SUPER EFFICENT SOLITION: n= 94 Product= 643805382
6. SORTING SOLUTION: n= 94 Product: 643805382
7. SUPER EFFICENT SOLITION: n= 40 Product= 545079705
7. SORTING SOLUTION: n= 40 Product: 545079705
8. SUPER EFFICENT SOLITION: n= 81 Product= 646703850
8. SORTING SOLUTION: n= 81 Product: 646703850
9. SUPER EFFICENT SOLITION: n= 63 Product= 642648552
9. SORTING SOLUTION: n= 63 Product: 642648552
10. SUPER EFFICENT SOLITION: n= 29 Product= 626954166
10. SORTING SOLUTION: n= 29 Product: 626954166
11. SUPER 